extensions/sockets/src/main/java/org/mule/module/socket/api/connection/udp/UdpListenerConnection.java

```
@@ -0,0 +1,90 @@
+     * for a response, one will be sent but not from the same listener socket the source has. The response
+     * will be sent from a new different {@link DatagramSocket} bound to a port choose by the system.
+     * <p>
+     * The new {@link DatagramSocket} used for responding has the same configuration settings than the
+     * listener socket used in the {@link Source}.
+     *
+     * @throws ReadingTimeoutException if the socket reached timeout while awaiting the arrival of a new package.
+     * @throws ConnectionException     if the socket was close while blocked on {@link DatagramSocket#receive(DatagramPacket)} method
+     *                                 and the close was not intentionally done by the {@link UdpListenerConnection#disconnect()} method.
+     * @throws IOException             if the socket was close while blocked on {@link DatagramSocket#receive(DatagramPacket)} method.
+     */
+    @Override
+    public SocketWorker listen(MuleContext muleContext, MessageHandler<InputStream, SocketAttributes> messageHandler) throws IOException, ConnectionException
+    {
+        DatagramPacket packet = createPacket(socketProperties.getReceiveBufferSize());
+
+        try
+        {
+            socket.receive(packet);
+        }
+        catch (SocketTimeoutException e)
+        {
+            throw new ReadingTimeoutException("UDP Source timed out while awaiting for new packages", e);
+        }
+        catch (IOException e)
+        {
+            if (!wasDisconnected)
+            {
+                throw new ConnectionException("An error occurred while listening for new UDP packets", e);
+            }
+
+            // socket was not intentionally closed by the connection provider
+            if (LOGGER.isDebugEnabled())
+            {
+                LOGGER.debug("UDP listener socket has been gracefully closed");
+            }
+
+            throw e;
```

if it was gracefully closed then why the need to propagate the exception?